src/analyzing/CFG_build.ml

```diff
@@ -240,41 +259,122 @@ let rec cfg_stmt : state -> F.nodei option -> stmt -> cfg_stmt_result =
   | Return (tok, e) ->
       let newi = state.g#add_node { F.n = F.NReturn (tok, e) } in
       state.g |> add_arc_from_opt (previ, newi);
-      state.g |> add_arc (newi, state.exiti);
+      state.g |> add_arc (newi, state.return_destination);
+      state.may_return := true;
       CfgFirstLast (newi, None)
-  | Try (try_st, catches, finally_st) ->
+  | Try (try_st, catches, else_st, finally_st) ->
       (* previ ->
        * newi ->
        * try -> catchesi --> catch1 -|
        *                 |->  ...   -|
        *                 |-> catchN -|
-       *                 |-----------|-> newfakefinally -> finally
-       *
+       *                 |-----------|
+       *     -> elsei    --> else    |-> newfakefinally -> finally
        *)
       let newi = state.g#add_node { F.n = NOther (Noop "try") } in
-      state.g |> add_arc_from_opt (previ, newi);
       let catchesi = state.g#add_node { F.n = NOther (Noop "catch") } in
-      let state' = { state with try_catches_opt = Some catchesi } in
-      let finaltry = cfg_stmt_list state' (Some newi) try_st in
-      state.g |> add_arc_from_opt (finaltry, catchesi);
+      let elsei = state.g#add_node { F.n = NOther (Noop "else") } in
       let newfakefinally = state.g#add_node { F.n = NOther (Noop "finally") } in
+
+      (* In the try/except/else clauses, if there is a return, then
+       * execute the finally clause, or if there's no finally clause, propagate
+       * the return.
+       *)
+      let return_destination =
+        match finally_st with
+        | [] -> state.return_destination
+        | _ -> newfakefinally
+      in
+
+      (* From prev to try. *)
+      state.g |> add_arc_from_opt (previ, newi);
+
+      (* Inside try may go to catches. *)
+      let try_state =
+        { state with throw_destination = Some catchesi; return_destination }
+      in
+      let finaltry = cfg_stmt_list try_state (Some newi) try_st in
+
+      (* The following will help determine whether we should exit
+       * or continue the flow after the whole statement
+       *)
+      (* When this is true, the flow stops after the finally clause. *)
+      let try_must_return =
+        match finaltry with
+        | None -> !(state.may_return) && not !(state.may_throw)
+        | _ -> false
+      in
+      (* When this is true, control must go to the else clause. *)
+      let try_must_not_exit =
+        match finaltry with
+        | None -> false
+        | _ -> (not !(state.may_throw)) && not !(state.may_return)
```

Comment:
Better `Some _`. In general we consider `_` wrong unless enumerating all possibilities is unreasonable.

Reply:
Good to know!